diff --git a/dlls/d2d1/wic_render_target.c b/dlls/d2d1/wic_render_target.c
index 858c187e3bc..77d2adb705e 100644
--- a/dlls/d2d1/wic_render_target.c
+++ b/dlls/d2d1/wic_render_target.c
@@ -197,6 +197,7 @@ HRESULT d2d_wic_render_target_init(struct d2d_wic_render_target *render_target,
     switch (texture_desc.Format)
     {
         case DXGI_FORMAT_B8G8R8A8_UNORM:
+        case DXGI_FORMAT_R8G8B8A8_UNORM:
             render_target->bpp = 4;
             break;
 
