From 8c4b7160e01ba89a81d44a1ece982e311a2364bf Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Wed, 12 Feb 2020 13:13:52 +0100
Subject: Disable the compositor bypass


diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 7bf9182705..e81f0625d9 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1123,7 +1123,7 @@ void update_net_wm_states( struct x11drv_win_data *data )
             new_state |= (1 << NET_WM_STATE_MAXIMIZED);
         else if (!(style & WS_MINIMIZE))
 	{
-            net_wm_bypass_compositor = 1;
+            net_wm_bypass_compositor = 0;
             if (!wm_is_steamcompmgr(data->display) || !fs_hack_enabled())
                 /* when fs hack is enabled, we don't want steamcompmgr to resize the window to be fullscreened */
                 new_state |= (1 << NET_WM_STATE_FULLSCREEN);
