--- a/dlls/ntdll/unix/loader.c	2025-02-23 02:31:43.695646158 +0500
+++ b/dlls/ntdll/unix/loader.c	2025-02-23 04:17:16.924362254 +0500
@@ -542,7 +542,7 @@
 char *get_alternate_wineloader( WORD machine )
 {
     const char *arch;
-    BOOL force_wow64 = (arch = getenv( "WINEARCH" )) && !strcmp( arch, "wow64" );
+    BOOL force_wow64 = !(arch = getenv( "WINEARCH" )) || !strcmp( arch, "wow64" );
     char *ret = NULL;
 
     if (is_win64)
