From d37b7f069fa019fbcc8dd51af836ce41b2d59a87 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Thu, 11 Mar 2021 16:39:36 -0600
Subject: [PATCH 03/31] autoreconf

---
 configure           | 6 ++++++
 include/config.h.in | 3 +++
 2 files changed, 9 insertions(+)

diff --git a/configure b/configure
index 5948756e7ae..1b5a543c000 100755
--- a/configure
+++ b/configure
@@ -8365,6 +8365,12 @@ if test "x$ac_cv_header_linux_ucdrom_h" = xyes
 then :
   printf "%s\n" "#define HAVE_LINUX_UCDROM_H 1" >>confdefs.h
 
+fi
+ac_fn_c_check_header_compile "$LINENO" "linux/winesync.h" "ac_cv_header_linux_winesync_h" "$ac_includes_default"
+if test "x$ac_cv_header_linux_winesync_h" = xyes
+then :
+  printf "%s\n" "#define HAVE_LINUX_WINESYNC_H 1" >>confdefs.h
+
 fi
 ac_fn_c_check_header_compile "$LINENO" "lwp.h" "ac_cv_header_lwp_h" "$ac_includes_default"
 if test "x$ac_cv_header_lwp_h" = xyes
diff --git a/include/config.h.in b/include/config.h.in
index a0165765a85..2f9085ef7f2 100644
--- a/include/config.h.in
+++ b/include/config.h.in
@@ -275,6 +275,9 @@
 /* Define to 1 if you have the <linux/videodev2.h> header file. */
 #undef HAVE_LINUX_VIDEODEV2_H
 
+/* Define to 1 if you have the <linux/winesync.h> header file. */
+#undef HAVE_LINUX_WINESYNC_H
+
 /* Define to 1 if you have the <lwp.h> header file. */
 #undef HAVE_LWP_H
 
-- 
2.34.1

