From 636c89d36a03a42508f63ac2ddb5a7850a0b413a Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Sun, 15 Sep 2019 14:25:12 +0200
Subject: Revert winex11.drv: Wake up the display on user input.

Signed-off-by: Zebediah Figura <zfigura@codeweavers.com>
Signed-off-by: Alexandre Julliard <julliard@winehq.org>

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 9ea5ff0f93..7b2cdc68a6 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -1909,22 +1909,6 @@ BOOL CDECL X11DRV_CreateDesktopWindow( HWND hwnd )
 }
 
 
-static WNDPROC desktop_orig_wndproc;
-
-#define WM_WINE_NOTIFY_ACTIVITY WM_USER
-
-static LRESULT CALLBACK desktop_wndproc_wrapper( HWND hwnd, UINT msg, WPARAM wp, LPARAM lp )
-{
-    switch (msg)
-    {
-    case WM_WINE_NOTIFY_ACTIVITY:
-        XResetScreenSaver( gdi_display );
-        XFlush( gdi_display );
-        break;
-    }
-    return desktop_orig_wndproc( hwnd, msg, wp, lp );
-}
-
 /**********************************************************************
  *		CreateWindow   (X11DRV.@)
  */
@@ -1935,9 +1919,6 @@ BOOL CDECL X11DRV_CreateWindow( HWND hwnd )
         struct x11drv_thread_data *data = x11drv_init_thread_data();
         XSetWindowAttributes attr;
 
-        desktop_orig_wndproc = (WNDPROC)SetWindowLongPtrW( hwnd, GWLP_WNDPROC,
-                                                           (LONG_PTR)desktop_wndproc_wrapper );
-
         /* create the cursor clipping window */
         attr.override_redirect = TRUE;
         attr.event_mask = StructureNotifyMask | FocusChangeMask;
