From 9b588663df7b2e38bb03db4e67f7b9274d62eee7 Mon Sep 17 00:00:00 2001
From: Paul Gofman <gofmanp@gmail.com>
Date: Fri, 16 Aug 2019 16:19:15 +0300
Subject: [PATCH] kernelbase: Remove DECLSPEC_HOTPATCH from
 SetThreadStackGuarantee().

Signed-off-by: Paul Gofman <gofmanp@gmail.com>
---
 dlls/kernelbase/thread.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/kernelbase/thread.c b/dlls/kernelbase/thread.c
index cf8e6be2b2..fb69276f84 100644
--- a/dlls/kernelbase/thread.c
+++ b/dlls/kernelbase/thread.c
@@ -468,7 +468,7 @@ BOOL WINAPI DECLSPEC_HOTPATCH SetThreadPriorityBoost( HANDLE thread, BOOL disabl
 /**********************************************************************
  *           SetThreadStackGuarantee   (kernelbase.@)
  */
-BOOL WINAPI DECLSPEC_HOTPATCH SetThreadStackGuarantee( ULONG *size )
+BOOL WINAPI SetThreadStackGuarantee( ULONG *size )
 {
     static int once;
     if (once++ == 0) FIXME("(%p): stub\n", size);
-- 
2.21.0

