From aa73d7d11fb4897de8f931051ed98858cf73ac69 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 2 Apr 2020 20:41:43 +0200
Subject: [PATCH] ntdll: Guard seccomp with WINESECCOMP env var.

---
 dlls/ntdll/signal_x86_64.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/ntdll/signal_x86_64.c b/dlls/ntdll/signal_x86_64.c
index dd9c5e8c77c..9ffce27b515 100644
--- a/dlls/ntdll/signal_x86_64.c
+++ b/dlls/ntdll/signal_x86_64.c
@@ -3342,9 +3342,18 @@ static void install_bpf(struct sigaction *sig_act)
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_TRAP),
        BPF_STMT(BPF_RET | BPF_K, SECCOMP_RET_ALLOW),
     };
+    static int enable_seccomp = -1;
     struct sock_fprog prog;
     int ret;
 
+    if (enable_seccomp == -1)
+        enable_seccomp = getenv("WINESECCOMP") && atoi(getenv("WINESECCOMP"));
+
+    if (!enable_seccomp)
+        return;
+
+    MESSAGE("wine: enabling seccomp syscall filters.\n");
+
     memset(&prog, 0, sizeof(prog));
     prog.len = ARRAY_SIZE(filter);
     prog.filter = filter;
