--- a/dlls/d3d9/d3d9_main.c	
+++ a/dlls/d3d9/d3d9_main.c	
@@ -76,6 +76,62 @@ HRESULT WINAPI DECLSPEC_HOTPATCH Direct3DCreate9Ex(UINT sdk_version, IDirect3D9E
     return D3D_OK;
 }
 
+static int uisv_refcount;
+
+static HRESULT WINAPI uisv_QueryInterface(void *iface, REFIID riid, void **out)
+{
+    return E_NOINTERFACE;
+}
+
+static ULONG WINAPI uisv_AddRef(void *iface)
+{
+    ULONG refcount = InterlockedIncrement(&uisv_refcount);
+    return refcount;
+}
+
+static ULONG WINAPI uisv_Release(void *iface)
+{
+    ULONG refcount = InterlockedDecrement(&uisv_refcount);
+    return refcount;
+}
+
+static HRESULT WINAPI uisv_DefaultUnimp()
+{
+    FIXME("uisv_DefaultUnimp() called: will crash\n");
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI uisv_1par(DWORD p1)
+{
+    FIXME("uisv_4par called\n");
+    return S_OK;
+}
+
+static HRESULT WINAPI uisv_4par(DWORD p1, DWORD p2, DWORD p3, DWORD p4)
+{
+    FIXME("uisv_4par called\n");
+    return S_OK;
+}
+
+static HRESULT WINAPI uisv_5par(DWORD p1, DWORD p2, DWORD p3, DWORD p4, DWORD p5)
+{
+    FIXME("uisv_5par called\n");
+    return S_OK;
+}
+
+
+static const void *undefinedInternalShaderValidator_vtbl[] =
+{
+    /* IUnknown */
+    uisv_QueryInterface,
+    uisv_AddRef,
+    uisv_Release,
+    /* interface */
+    uisv_4par,
+    uisv_5par,
+    uisv_1par
+ };
+
 /*******************************************************************
  *       Direct3DShaderValidatorCreate9 (D3D9.@)
  *
@@ -85,9 +141,9 @@ HRESULT WINAPI DECLSPEC_HOTPATCH Direct3DCreate9Ex(UINT sdk_version, IDirect3D9E
 void* WINAPI Direct3DShaderValidatorCreate9(void)
 {
     static int once;
-
+    static void *vtbl_addr=undefinedInternalShaderValidator_vtbl;
     if (!once++) FIXME("stub\n");
-    return NULL;
+    return &vtbl_addr;
 }
 
 /***********************************************************************
