From f68fbcf73179395e88113e8c240e2fadcbb3b7f0 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Sun, 19 Oct 2025 15:02:36 +0200
Subject: container: Move to latest proton OWC/umu-sdk


diff --git a/Makefile.in b/Makefile.in
index 92d78ac0..ec6f588f 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -24,7 +24,7 @@ SHELL := /bin/bash
 #   BUILD_NAME      - Name of the build for manifests etc.
 #   STEAMRT_IMAGE   - Name of the docker image to use for building
 
-STEAMRT_IMAGE ?= registry.gitlab.steamos.cloud/proton/sniper/sdk:3.0.20250210.116596-0
+STEAMRT_IMAGE ?= ghcr.io/open-wine-components/umu-sdk:latest
 ifeq ($(SRCDIR),)
 	foo := $(error SRCDIR not set, do not include Makefile.in directly, run ./configure.sh to generate Makefile)
 endif
