From a5e0cc29b018879910556378f2b5ec5be2354847 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Tue, 22 Jul 2025 16:50:07 +0200
Subject: Disable autoconf werror for wine


diff --git a/Makefile.in b/Makefile.in
index 4d1fa3be..531ee321 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -502,7 +502,6 @@ WINE_SOURCE_ARGS = \
   --exclude dlls/winevulkan/winevulkan.spec
 
 WINE_CONFIGURE_ARGS = \
-  --enable-werror \
   --with-mingw \
   --disable-tests
 
