From 1b6e416742d381d786c88603046a93feec261808 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Sun, 22 Mar 2020 02:08:35 +0100
Subject: Remove d3d10 and d3d10_1 for dxvk 1.6 and wine 5.3+


diff --git a/proton b/proton
index be3586d..2d8a659 100755
--- a/proton
+++ b/proton
@@ -333,7 +333,7 @@ class CompatData:
                         os.remove(dstfile)
                     try_copy(srcfile, dstfile)
 
-            dxvkfiles = ("d3d11", "d3d10", "d3d10core", "d3d10_1")
+            dxvkfiles = ("d3d11", "d3d10core")
             def make_dxvk_links(dll_dir, link_dir):
                 for f in dxvkfiles:
                     dst = link_dir + "/" + f + ".dll"
