From f319dc029be11156c22429c8e0892b4c0e3d0657 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Mon, 23 Mar 2020 22:24:54 +0100
Subject: dxvk_config lib support


diff --git a/proton b/proton
index be3586d..75e2e93 100755
--- a/proton
+++ b/proton
@@ -333,7 +333,7 @@ class CompatData:
                         os.remove(dstfile)
                     try_copy(srcfile, dstfile)
 
-            dxvkfiles = ("d3d11", "d3d10core")
+            dxvkfiles = ("dxvk_config", "d3d11", "d3d10core")
             def make_dxvk_links(dll_dir, link_dir):
                 for f in dxvkfiles:
                     dst = link_dir + "/" + f + ".dll"
