From f319dc029be11156c22429c8e0892b4c0e3d0657 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Mon, 23 Mar 2020 22:24:54 +0100
Subject: dxvk_config lib support


diff --git a/proton b/proton
index be3586d..75e2e93 100755
--- a/proton
+++ b/proton
@@ -333,7 +333,7 @@ class CompatData:
                 dxvkfiles = ["dxvk_config"]
                 wined3dfiles = ["d3d11", "d3d10", "d3d10core", "d3d10_1", "d3d9"]
             else:
-                dxvkfiles = ["d3d11", "d3d10core", "d3d9"]
+                dxvkfiles = ["dxvk_config", "d3d11", "d3d10core", "d3d9"]
                 wined3dfiles = []
 
             #if the user asked for dxvk's dxgi (dxgi=n), then copy it into place
