From 68621c9d55a19a6820b3f4ca3e9fb1a2464412c5 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Tue, 11 Feb 2020 11:23:10 +0100
Subject: proton-tkg: makepkg build adjustments


diff --git a/proton b/proton
index 811de5e..7aacb34 100755
--- a/proton
+++ b/proton
@@ -103,7 +103,7 @@ class Proton:
         self.wine_bin = self.bin_dir + "wine"
         self.wineserver_bin = self.bin_dir + "wineserver"
         self.winetricks_bin = self.base_dir + "/dist/share/winetricks/winetricks_runner"
-        self.dist_lock = FileLock(self.path("dist.lock"), timeout=-1)
+        self.dist_lock = FileLock("/tmp/dist.lock", timeout=-1)
 
     def path(self, d):
         return self.base_dir + d
@@ -401,15 +401,6 @@ def setup_prefix(self):
                 makedirs(self.prefix_dir + "/drive_c")
                 set_dir_casefold_bit(self.prefix_dir + "/drive_c")
 
-            if not os.path.exists(self.prefix_dir + "/user.reg"):
-                self.copy_pfx()
-
-            if not os.path.lexists(self.prefix_dir + "/dosdevices/c:"):
-                os.symlink("../drive_c", self.prefix_dir + "/dosdevices/c:")
-
-            if not os.path.lexists(self.prefix_dir + "/dosdevices/z:"):
-                os.symlink("/", self.prefix_dir + "/dosdevices/z:")
-
             # collect configuration info
             if "STEAM_COMPAT_CLIENT_INSTALL_PATH" in os.environ:
                 #modern steam client sets this
@@ -789,6 +789,7 @@ class Session:
         if local_env is None:
             local_env = self.env
         subprocess.call(args, env=local_env, stderr=self.log_file, stdout=self.log_file)
+        self.env["PYTHONPYCACHEPREFIX"] = "/tmp"
 
     def run(self):
         if "PROTON_DUMP_DEBUG_COMMANDS" in self.env and nonzero(self.env["PROTON_DUMP_DEBUG_COMMANDS"]):
@@ -1208,7 +1208,6 @@ if __name__ == "__main__":
     g_proton = Proton(os.path.dirname(sys.argv[0]))
 
     g_proton.cleanup_legacy_dist()
-    g_proton.do_steampipe_fixups()
 
     g_compatdata = CompatData(os.environ["STEAM_COMPAT_DATA_PATH"])
 
@@ -1217,8 +1217,6 @@ if __name__ == "__main__":
 
     g_session.init_wine()
 
-    if g_proton.missing_default_prefix():
-        g_proton.make_default_prefix()
 
     g_session.init_session()
 
