From 68621c9d55a19a6820b3f4ca3e9fb1a2464412c5 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Tue, 11 Feb 2020 11:23:10 +0100
Subject: proton-tkg: makepkg build adjustments


diff --git a/proton b/proton
index 811de5e..7aacb34 100755
--- a/proton
+++ b/proton
@@ -103,7 +103,7 @@ class Proton:
         self.wine_bin = self.bin_dir + "wine"
         self.wineserver_bin = self.bin_dir + "wineserver"
         self.winetricks_bin = self.base_dir + "/dist/share/winetricks/winetricks_runner"
-        self.dist_lock = FileLock(self.path("dist.lock"), timeout=-1)
+        self.dist_lock = FileLock("/tmp/dist.lock", timeout=-1)
 
     def path(self, d):
         return self.base_dir + d
@@ -175,7 +175,7 @@ class Proton:
 
     def do_steampipe_fixups(self):
         fixups_json = self.path("steampipe_fixups.json")
-        fixups_mtime = self.path("files/steampipe_fixups_mtime")
+        fixups_mtime = ("/tmp/steampipe_fixups_mtime")
 
         if os.path.exists(fixups_json):
             with self.dist_lock:
@@ -789,6 +789,7 @@ class Session:
         if local_env is None:
             local_env = self.env
         subprocess.call(args, env=local_env, stderr=self.log_file, stdout=self.log_file)
+        self.env["PYTHONPYCACHEPREFIX"] = "/tmp"
 
     def run(self):
         if "PROTON_DUMP_DEBUG_COMMANDS" in self.env and nonzero(self.env["PROTON_DUMP_DEBUG_COMMANDS"]):
