From 84cbe52bdacc382f8addd76d3a4492e199ba3fac Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Fri, 9 Aug 2019 22:07:37 +0200
Subject: Use Steam helper


diff --git a/proton b/proton
index c0857e793..e305148b6 100755
--- a/proton
+++ b/proton
@@ -494,6 +494,7 @@ class Session:
         self.log_file = None
         self.env = dict(os.environ)
         self.dlloverrides = {
+                "steam.exe": "b", #always use our special built-in steam.exe
                 "mfplay": "n", #disable built-in mfplay
                 "dotnetfx35.exe": "b,n" #replace the broken installer, as does Windows, if it exists
         }
@@ -761,7 +762,7 @@ class Session:
                     f.write(" \"" + arg + "\"")
             f.write(")\n")
             self.dump_dbg_env(f)
-            f.write("\t\"" + g_proton.wine_bin + "\" \"${@:-${DEF_CMD[@]}}\"\n")
+            f.write("\t\"" + g_proton.wine_bin + "\" steam.exe \"${@:-${DEF_CMD[@]}}\"\n")
         os.chmod(tmpdir + "run", 0o755)
 
     def run_proc(self, args, local_env=None):
@@ -990,9 +990,9 @@ class Session:
         ]
 
         if any(check_args) and not "alternativestart" in self.compat_config:
-            commandtorun=[g_proton.wine_bin] + sys.argv[2:] + self.cmdlineappend
+            commandtorun=[g_proton.wine_bin, "steam"] + sys.argv[2:] + self.cmdlineappend
         else:
-            commandtorun=[g_proton.wine_bin] + sys.argv[2:] + self.cmdlineappend
+            commandtorun=[g_proton.wine_bin, "start.exe", "/b", "steam"] + sys.argv[2:] + self.cmdlineappend
 
         if "winetricks" in self.compat_config:
             if os.path.exists(os.environ["STEAM_COMPAT_DATA_PATH"]):
@@ -1015,9 +1015,9 @@ class Session:
                                     open(tricksfile_path, "w+").close()
                     except ImportError:
                         if any(check_args) and not "alternativestart" in self.compat_config:
-                            commandtorun=[g_proton.wine_bin] + sys.argv[2:] + self.cmdlineappend
+                            commandtorun=[g_proton.wine_bin, "steam"] + sys.argv[2:] + self.cmdlineappend
                         else:
-                            commandtorun=[g_proton.wine_bin] + sys.argv[2:] + self.cmdlineappend
+                            commandtorun=[g_proton.wine_bin, "start.exe", "/b", "steam"] + sys.argv[2:] + self.cmdlineappend
         self.run_proc(commandtorun)
 
 
