From 802b9f17837446baf7a4d524b6b843ef30f62347 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Thu, 24 Nov 2022 14:55:50 +0100
Subject: Fixup for ce91ef6426bf5065bd31bb82fa4f76011e7a9a36


diff --git a/steam_helper/steam.cpp b/steam_helper/steam.cpp
index eca32f5..c044580 100644
--- a/steam_helper/steam.cpp
+++ b/steam_helper/steam.cpp
@@ -1446,7 +1446,7 @@ int main(int argc, char *argv[])
     }
 
     if (game_process)
-        NtSetInformationProcess( GetCurrentProcess(), (PROCESS_INFORMATION_CLASS)1000 /* ProcessWineMakeProcessSystem */,
+        NtSetInformationProcess( GetCurrentProcess(), (PROCESSINFOCLASS)1000 /* ProcessWineMakeProcessSystem */,
                                  &wait_handle, sizeof(HANDLE *) );
 
     if(wait_handle != INVALID_HANDLE_VALUE)
