From 39105fd384e65f5dc2951d9d065d7fc7bdfc7c7d Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Wed, 8 Apr 2020 01:21:47 +0200
Subject: lsteamclient: Get rid of wine/library.h dep


diff --git a/lsteamclient/steamclient_main.c b/lsteamclient/steamclient_main.c
index de54021..d952d2d 100644
--- a/lsteamclient/steamclient_main.c
+++ b/lsteamclient/steamclient_main.c
@@ -14,7 +14,6 @@
 #include "winbase.h"
 #include "winnls.h"
 #include "wine/debug.h"
-#include "wine/library.h"
 #include "wine/list.h"
 #include "steam_defs.h"
 
