From 5c5a18438870d10c31933061b2721477658931c9 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Mon, 3 Feb 2020 07:17:25 +0100
Subject: proton-tkg: initial vr support


diff --git a/proton b/proton
index 811de5e..028e7dd 100755
--- a/proton
+++ b/proton
@@ -333,6 +333,15 @@ class CompatData:
                         os.remove(dstfile)
                     try_copy(steamdir + "/legacycompat/" + f, dstfile)
 
+            #copy openvr files into place
+            dst = self.prefix_dir + "/drive_c/vrclient/bin/"
+            makedirs(dst)
+            try_copy(g_proton.lib_dir + "wine/fakedlls/vrclient.dll", dst)
+            try_copy(g_proton.lib64_dir + "wine/fakedlls/vrclient_x64.dll", dst)
+
+            try_copy(g_proton.lib_dir + "wine/dxvk/openvr_api_dxvk.dll", self.prefix_dir + "/drive_c/windows/syswow64/")
+            try_copy(g_proton.lib64_dir + "wine/dxvk/openvr_api_dxvk.dll", self.prefix_dir + "/drive_c/windows/system32/")
+
             dxvkfiles = ("d3d11", "d3d10", "d3d10core", "d3d10_1")
             def make_dxvk_links(dll_dir, link_dir):
                 for f in dxvkfiles:
@@ -671,9 +671,6 @@ class Session:
         if "noconhost" in self.compat_config:
             self.dlloverrides["conhost.exe"] = "d"
 
-        # Disable openvr lib as we don't support it
-        self.dlloverrides["openvr_api_dxvk"] = "d"
-
         s = ""
         for dll in self.dlloverrides:
             setting = self.dlloverrides[dll]
