From 802436bbef59b1d5ff17a8fbd4d5fc0294cb3525 Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Wed, 8 Apr 2020 05:33:48 +0200
Subject: vrclient: Get rid of wine/library.h dep


diff --git a/vrclient_x64/vrclient_x64/vrclient_main.c b/vrclient_x64/vrclient_x64/vrclient_main.c
index ea8e375..cf5888f 100644
--- a/vrclient_x64/vrclient_x64/vrclient_main.c
+++ b/vrclient_x64/vrclient_x64/vrclient_main.c
@@ -9,7 +9,6 @@
 #include "winbase.h"
 #include "winnls.h"
 #include "wine/debug.h"
-#include "wine/library.h"
 
 #include "vrclient_defs.h"
 #include "vrclient_private.h"
