From 63a91cb7793723e5b9aca11da179736f6d05c70e Mon Sep 17 00:00:00 2001
From: Tk-Glitch <ti3nou@gmail.com>
Date: Thu, 2 Aug 2018 21:07:33 +0200
Subject: Low latency alsa patch stolen from https://blog.thepoon.fr/osuLinuxAudioLatency/

https://aur.archlinux.org/packages/wine-osu/

diff --git a/dlls/winealsa.drv/alsa.c b/dlls/winealsa.drv/alsa.c
index b285705509..ec90fa1d11 100644
--- a/dlls/winealsa.drv/alsa.c
+++ b/dlls/winealsa.drv/alsa.c
@@ -81,10 +81,10 @@
     pthread_mutex_t lock;
 };

-#define EXTRA_SAFE_RT 40000
+#define EXTRA_SAFE_RT 4000

-static const REFERENCE_TIME def_period = 100000;
-static const REFERENCE_TIME min_period = 50000;
+static const REFERENCE_TIME def_period = 4000;
+static const REFERENCE_TIME min_period = 2000;
 
 static NTSTATUS alsa_not_implemented(void *args)
 {
 
